src/stats.rs

```diff
@@ -686,65 +885,145 @@ impl Stat {
             .collect::<Vec<_>>()
     }
 
-    // Get all the grouped value names currently tracked.
-    fn get_group_name_vals(&self) -> Vec<(String, f64)> {
-        // Only hold the read lock long enough to get the keys and values.
-        let inner_vals = self.group_values.read().expect("Poisoned lock)");
-        inner_vals
-            .iter()
-            .map(|(k, v)| (k.clone(), v.as_float()))
-            .collect()
+    /// Get all the grouped/bucketed value names currently tracked.
+    fn get_bucket_group_name_vals(&self) -> Vec<(Option<String>, Option<usize>, f64)> {
+        let values = if let Some(_) = self.buckets {
+            if self.is_grouped {
+                let mut bucket_group_vals = Vec::new();
+                let inner_vals = self.bucket_group_values.read().expect("Poisoned lock)");
+                for (group_values_str, bucket_vals) in inner_vals.iter() {
+                    bucket_group_vals.extend(bucket_vals.iter().enumerate().map(|(i, val)| {
+                        (Some(group_values_str.to_string()), Some(i), val.as_float())
+                    }));
+                }
+                bucket_group_vals
+            } else {
+                self.bucket_values
+                    .iter()
+                    .enumerate()
+                    .map(|(index, value)| (None, Some(index), value.as_float()))
+                    .collect()
+            }
+        } else {
+            if self.is_grouped {
+                // Only hold the read lock long enough to get the keys and values.
+                let inner_vals = self.group_values.read().expect("Poisoned lock)");
+                inner_vals
+                    .iter()
+                    .map(|(group_values_str, value)| {
+                        (Some(group_values_str.to_string()), None, value.as_float())
+                    })
+                    .collect()
+            } else {
+                vec![(None, None, self.value.as_float())]
+            }
+        };
+        values
     }
 
-    // Check if the stat is grouped, and if so, get and update the correct
-    // grouped values.
-    fn check_update_grouped(&self, defn: &StatDefinition, trigger: &StatTrigger) {
+    /// Update the stat's value(s) according to the given `StatTrigger` and `StatDefinition`.
+    fn update(&self, defn: &StatDefinition, trigger: &StatTrigger) {
```

Comment:
This method is really hard to read and parse.  Could we refactor a bit and maybe split up subfunctions for handling groups, buckets, etc?